qiskit/circuit/quantumcircuit.py

```diff
@@ -2854,7 +2854,14 @@ def _assign_parameter(self, parameter: Parameter, value: ParameterValueType) ->
                     new_param = assignee.assign(parameter, value)
                     # if fully bound, validate
                     if len(new_param.parameters) == 0:
-                        instr.params[param_index] = instr.validate_parameter(new_param)
+                        if new_param._symbol_expr.is_integer and new_param.is_real():
+                            val = int(new_param)
+                        elif new_param.is_real():
+                            # Workaround symengine not supporting float(<ComplexDouble>)
+                            val = complex(new_param).real
+                        else:
+                            val = complex(new_param)
```

Comment:
Just checking that we expect complex parameters possibly here? I didn't allow them for the calibration keys below. Perhaps I should? The previous code did not allow it and support for complex parameters in pulse has been deprecated in https://github.com/Qiskit/qiskit-terra/pull/9897 so I think it is not needed.

Reply:
I don't think complex parameters are possible here. At least historically complex parameters were only expected in a pulse schedule. But @Cryoris can confirm (since he originally added support for complex parameters).

Reply:
Yeah that's right, so far the only use case was the pulses. If they are deprecated there, I don't think there is any other part that requires supporting binding complex parameter values.

Reply:
I took the complex case out but put it back just now because a couple tests were failing. The tests were trying to assign a complex value and see that an appropriate exception was raised by the `validate_parameter` method of the instruction. We can leave it for later whether we want to lift up the exception raising to a higher level of than individual instructions' `validate_parameter`.